### DISCUSSION TITLE
Evaluation FerretDB

### DISCUSSION THREAD

**Discussion by decebals-btcs** [FerretDB/FerretDB#4161] [General]
First of all, I would like to start by thanking you for this initiative/idea.

I am trying to evaluate FerretDB in order to migrate from MongoDB.
I have some questions, uncertainties and problems that I will share in the following:
- is it production ready, does anyone use it in production? if the answer is yes, is there somewhere a link with the applications/companies that use FerretDB?
- I see from https://github.com/FerretDB/FerretDB/issues/175 that the part with change streams is not implemented yet, there is an open issue on this topic, but is it known if it will be implemented and when? or does it provide another change feed implementation mechanism?
I think that the main purpose of FerretDB is to implement part of the functionalities offered by MongoDB, but I wonder if it is desired to add new capabilities, specific to FerretDB, in the form of new commands.
- I know that FerretDB can import data from MongoDB (via dump), but is there a possibility to export data from FerretDB and import it into MongoDB (it's somehow like a safety net)?
- in terms of transactionality, I see from https://github.com/FerretDB/FerretDB/issues/153 that Session commands are not implemented (for example startSession, commitTransaction, abortTransaction). Is there a plan to implement them or does FerretDB offers another mechanism to solve this problem (transactionality)?
- related to production use, what scaling recommendations do you have in terms of clustering, HA, Failover, and so one?
- is there any builtin mechanism related to references? (in context of https://www.mongodb.com/docs/manual/reference/database-references)
- is there any mechanism to enforce a json schema validation mechanism for a MongoDB collection? (in context of https://github.com/FerretDB/FerretDB/issues/3070)

I know I asked a lot of questions and I thank you in advance for the answers!

By the way, I identified a problem in the evaluation process, the problem related to running queries from MongoDB Compass (I can connect to the database but cannot run queries). Probably in context of https://github.com/FerretDB/FerretDB/issues/1982. I attach a screenshot (MongoDB Compass version 1.42.2)

![image](https://github.com/FerretDB/FerretDB/assets/154612840/8ba538fc-ef1b-4cae-a04c-1f737f6d29be)

Good luck!

**Comment by decebals-btcs** [comment]
@AlekSi An opinion (answer) if possible?

**Comment by alwaqfi** [comment]
I would love to know the answer for these questions too. I'm currently working on a project and I'm between using FerretDB VS Postgres+Marten.

**Reply by rumyantseva** [reply]
@alwaqfi thank you for your interest! Since the original question was asked in March, we have some updates to share. To better answer your question, could you tell us more about your specific use case? What are your requirements and expectations regarding the database?

**Comment by leessmin** [comment]
I have an Intel Pentium series server, but the CPU does not support AVX, so I cannot install versions after MongoDB 5.0. I have to migrate to FerretDB. Once again, I would like to express my special gratitude to the FerretDB developers for their continuous contributions, which have provided me with a substitute for mongodb that I can us.

**Comment by AlekSi** [comment]
Yes, FerretDB is production-ready. One public example is FastNetMon: https://blog.ferretdb.io/fastnetmon-picked-ferretdb-saas-offering/. But, of course, you should evaluate FerretDB for your own use case.

Change streams will be supported eventually, but there is no target date yet. The currently supported mechanism is OpLog: https://docs.ferretdb.io/configuration/oplog-support/. Additional commands may be supported later.

Both mongodump/mongorestore and mongoexport/mongoimport should work for both MongoDB and FerretDB: https://docs.ferretdb.io/migration/migrating-from-mongodb/ Additionally, there is a proxy mode when data could be stored in both MongoDB and FerretDB: https://docs.ferretdb.io/migration/premigration-testing/

Yes, we plan to support transactions via sessions and other standard MongoDB APIs.

Currently, we don't provide any recommendations for clustering, HA, failover, etc.

References are purely driver-level features; they should already work with FerretDB.

JSON schema validation is not supported yet.

Finally, we are working on the next major FerretDB version, which is based on the PostgreSQL extension. It will provide much better compatibility and performance. It is not yet public but should be released as an open source in a few months.

(And that's the main reason for the long response – in that work, quite a few community discussions slipped through the cracks. We are sorry, and we will do better.)
